### DISCUSSION TITLE
Exception counters

### DISCUSSION THREAD

**Discussion by KJlt77** [TPC-Council/HammerDB#73] [Q&A] [answered]
I have noticed something like this in the HammerDB code: exception when some exception then rollback. So NOPM counts succeeded and failed operations with no difference.

Statistics about exceptions occurrence would be nice addition to the log. E.g. I would like to know witch databases or DB configurations are prone to deadlocks.

**Comment by sm-shaw** [comment, marked as answer]
This should not be the case that NOPM succeeds when the transaction rolls back. NOPM is extracted directly from the district table and is only increased by the NEWORD stored procedure when it commits successfully.  NOPM is not impacted by any of the other stored procedures.  On the other hand TPM will be increased whether the transaction succeeds or rolls back. In fact TPM is extracted from the  transaction records of the system tables of the relevant database and therefore will also show the transactions of any other workloads being run against the database at the same time. A good first sanity check is the expected ratio between TPM and NOPM if eg TPM is higher and NOPM is lower than expected then this indicates that transactions are not succeeding.  The reason both are shown is the TPM can be polled without impacting the performance of the workload whereas NOPM would, therefore NOPM is only measured at the start and end of the test. If you want information on deadlocks, commits, rollbacks etc this can come from the database performance tables. Currently database metrics are shown for Oracle, the plan is to add these type of metrics for all of the databases.

**Comment by sm-shaw** [comment]
Closed as no further update
